codice/ddf
# [2.21.x] DDF-5360 Adds the ability to go through OAuth to download a resource

#### What does this PR do?
Added support for the OAuth Resource Owner Password Credentials Grant
- https://github.com/codice/ddf/pull/5607
Adds the ability to go through OAuth to download a resource
- https://github.com/codice/ddf/pull/5769
- https://github.com/codice/ddf-ui/pull/83
Changed user ID based OAuth token storage to session ID to better handle the guest user
- https://github.com/codice/ddf/pull/5682
- https://github.com/codice/ddf-ui/pull/74

#### Who is reviewing it? 
@SmithJosh @bakejeyner 

#### Select relevant component teams: 
@codice/security 

#### Ask 2 committers to review/merge the PR and tag them here.
@emmberk
@garrettfreibott
@stustison 

#### How should this be tested?
Create two sources (S1 and S2) to use OAuth

- [x] Both sources should be available
- [x] Test querying S1
  - Should get a notification to log in
  - Sign in and re-query
  - Query should succeed with no errors
- [x] Test querying S2
  - Should get a notification to authorize the source
  - Authorize and re-query
  - Query should succeed with no errors
- [x] Test session-based storage works
  - After logging in and querying successfully, sign out of DDF and Keycloak
  - Sign back in as the same user and try querying
  - You should be prompted to log in to Keycloak again
- [x] Test querying as guest
  - "Sign in" to intrigue as guest
  - Query a source and sign in to Keycloak
  - "Sign in" to intrigue as guest in another browser or in incognito mode
  - The second guest should be prompted to sign in before querying a source
- [x] Test downloading one of the results you get back
- [x] Test downloading an old result
  - After querying a source, remove your oauth tokens by deleting all files under <DDF_HOME>/data/content/store/token
  - Try downloading one of the results you had
  - You should get redirected to sign in
- [x] Verify downloading a resource from an OAuth protected source works. Download using
  - the download button
  - the inspector tab
  - the action tab

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
